cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/worker/TaskWorkerTestUtil.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.worker;
+
+import com.google.common.util.concurrent.Service;
+import com.google.common.util.concurrent.Uninterruptibles;
+import org.apache.twill.common.Threads;
+import org.apache.twill.internal.ServiceListenerAdapter;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Common TaskWorker test utility functions
+ */
+public class TaskWorkerTestUtil {
+
+  /**
+   * Waits for the given task worker to finish
+   * @param taskWorker {@link TaskWorkerService}
+   */
+  static void waitForTaskWorkerToFinish(TaskWorkerService taskWorker) {
+    CompletableFuture<Service.State> future = new CompletableFuture<>();
+    taskWorker.addListener(new ServiceListenerAdapter() {
```

Comment:
This is unsafe and can cause flaky test. If the service is already started and completed, the listener won't be registering any state changes.

Reply:
Split to two methods to add a listener and another for completion. Add listener is called before starting the service from the test. Refactored older tests to follow the same.